### PR TITLE
fix(k8s): fix builtdns annotations

### LINKED ISSUE #2347: Rename "builtindns" and "builtindnsport" with consistent hyphenation
### Summary

Every other Kuma annotation that consists of multiple words separates those words with hyphens. These annotations ought to be named `kuma.io/builtin-dns` and `kuma.io/builtin-dns-port`.

Recommend we continue to accept both forms, but document the hyphenated one.

### LINKED ISSUE #4458: Kuma injector overrides annotation configuration with default values
### What happened?

I am trying to configure a pod with annotations, using this [documentation](https://kuma.io/docs/1.6.x/reference/kubernetes-annotations/#labels).
The Kuma injector don't apply them and use default configuration.

**How to reproduce?**

I use Kind cluster for development.

Create a cluster.
```bash
cat > cluster-config.yaml <<EOF
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
 - role: control-plane
 - role: worker
EOF

kind create cluster --config ./cluster-config.yaml
```

Install Kuma control plane.
```bash
kumactl install control-plane | kubectl apply -f -
```

Create Kuma meshed pod.
```bash
cat > pod.yaml <<EOF
apiVersion: v1
kind: Pod
metadata:
  name: hello-world
  labels:
    kuma.io/sidecar-injection: enabled
  annotations:
    kuma.io/builtindns: enabled
    kuma.io/builtindnsport: "15058"
spec:
  restartPolicy: Never
  containers:
    - name: hello-world
      image: strm/helloworld-http
      imagePullPolicy: IfNotPresent
      stdin: true
      tty: true
EOF

kubectl apply -f pod.yaml
```

Check pod annotations.
```bash
kubectl describe pod hello-world
```

**Expected output:**
```
Name:         hello-world
Namespace:    default
Priority:     0
Node:         kind-worker/172.18.0.3
Start Time:   Fri, 10 Jun 2022 14:41:35 +0700
Labels:       kuma.io/sidecar-injection=enabled
Annotations:  kuma.io/builtindns: enabled
              kuma.io/builtindnsport: 15058
              ...
```

**Real output**:
```
Name:         hello-world
Namespace:    default
Priority:     0
Node:         kind-worker/172.18.0.3
Start Time:   Fri, 10 Jun 2022 14:41:35 +0700
Labels:       kuma.io/sidecar-injection=enabled
Annotations:  kuma.io/builtindns: enabled
              kuma.io/builtindnsport: 15053
              ...
```

Pod annotation `kuma.io/builtindnsport` has a default value `15053`, not `15058` as I set in pod's configuration.

**Kuma version**
Kuma: 1.6.0

**Kuma configuration**
Default configuration file without changes.

[kuma-cp.conf.txt](https://github.com/kumahq/kuma/files/8876871/kuma-cp.conf.txt)

### PR DESCRIPTION
**Pull request by lahabana**
- Make override of the port work
- Fix from `kuma.io/builtindnsport` to `kuma.io/builtin-dns-port`
- Fix from `kuma.io/builtindns` to `kuma.io/builtin-dns`
- Rewrite `annotations.go` to make it easier to use defaults
- Provide a standard way to mark pod annotations as deprecated and to log about them

Fix #4458
Fix #2347

Signed-off-by: Charly Molter <charly.molter@konghq.com>

### CODE REVIEW

File: UPGRADE.md [hunk 1]
```diff
@@ -10,8 +10,10 @@ does not have any particular instructions.
 
 ### Helm
 
-Under `cni.image`, the default values for `repository` and `registry` have been
+- Under `cni.image`, the default values for `repository` and `registry` have been
 changed to agree with the other `image` values.
+- We are deprecating `kuma.io/builtindns` and `kuma.io/builtindnsport` annotations in favour or the clearer `kuma.io/builtin-dns` and `kuma.io/builtin-dns-port` the code is backward compatible
```
Comment (michaelbeaumont) [inline UPGRADE.md:15]: > the code is backward compatible

What does this mean? That no action is required but users _should_ migrate?

**Review by michaelbeaumont — APPROVED** [review]
Nice, LGTM!

File: UPGRADE.md (hunk 1 above)
Comment (lahabana) [reply UPGRADE.md:15]: Yes I guess this wording is wrong. Do you have an idea on how to rephrase?

File: UPGRADE.md [hunk 2]
```diff
@@ -10,8 +10,10 @@ does not have any particular instructions.
...
+but you should migrate (a warning is present on the log if you are using the deprecated version).
```
Comment (michaelbeaumont) [inline UPGRADE.md:16]:
```suggestion
- We are deprecating `kuma.io/builtindns` and `kuma.io/builtindnsport` annotations in favour of the clearer `kuma.io/builtin-dns` and `kuma.io/builtin-dns-port`. The behavior of the new annotations is unchanged but you should migrate (a warning is present on the log if you are using the deprecated version).
```

Like that maybe?

**Review by jakubdyszkiewicz — APPROVED** [review]

**Pull request merged**